Use generic signature for local variables with entry in LVTT

mgudemann opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Use generic signature for local variables with entry in LVTT

This uses the generic type information of the local variable type table for the
type of local variables. Before, no local variable was considered to a have
generic type.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 574de4e582c2233c4083cc0aadee07bd13fc3fae).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85130707

smowton commented on jbmc/src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -768,9 +768,16 @@ void java_bytecode_convert_methodt::setup_local_variables(
-    // TODO: might need changing once descriptor/signature issue is resolved
-    t=java_type_from_string(v.var.descriptor);
+
+    const std::string method_name = id2string(method_id);
```
const ref

smowton commented on jbmc/src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -768,9 +768,16 @@ void java_bytecode_convert_methodt::setup_local_variables(
-    t=java_type_from_string(v.var.descriptor);
+
+    const std::string method_name = id2string(method_id);
+    const size_t method_name_end = method_name.rfind(":(");
```
Should probably invariant that these are found

smowton reviewed (commented):
Tests situation for this?

mgudemann commented:
@smowton this is used/required in TG

smowton commented:
And its tests aren't portable here? As it stands, as with anything only used in TG but supported in the CBMC main repository, there's a danger we break it without realising

mgudemann commented:
@smowton I don't think this change is visible to put it into a regression test, I could probably do a unit test.
Effectively the change is transparent for any code that doesn't make use of generic types.

hannes-steffenhagen-diffblue commented on jbmc/src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -768,9 +768,20 @@ void java_bytecode_convert_methodt::setup_local_variables(
+    INVARIANT(
+      method_name_end != std::string::npos &&
+        class_name_end != std::string::npos,
+      "Java method name does not have the right format.");
```
:pick: 
Invariant string should describe why the invariant is true, not what happened when it failed - I.e. in this case it'd be something like "Always true because the syntax of Java method names looks like this".

Also, is it safe to assume that we won't have been fed a nonsensical class file or one with a weird new format here? (I'm guessing this isn't something that's likely to change, genuine question though)

mgudemann replied:
the current hypothesis is that we get a class file that is compiled from a java source by a java compiler
We have already had class files generated by compilers for other languages that did break some hypotheses.

smowton commented:
Can you see the result in `--show-symbol-table`, in the types of symbols generated for local variables, for example?

hannes-steffenhagen-diffblue commented on jbmc/src/java_bytecode/java_local_variable_table.cpp:
```diff
@@ -768,9 +768,20 @@ void java_bytecode_convert_methodt::setup_local_variables(
+      method_name_end != std::string::npos &&
+        class_name_end != std::string::npos,
+      "Java method name does not have the right format.");
+    const std::string class_name = method_name.substr(0, class_name_end);
```
Could this go into a separate function, something along the lines of `class_name_from_method_name`?

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 16d2572ab671d581d15bdf7cea0d4f6673fdb1e6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85430988
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

Commit:
Add unit test to check for generic local variables

The test checks that a local variable has the correct generic type and
instantiation.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 405fe5f425461604c04b7649cc23eeba40aa71b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85435436

mgudemann commented:
@smowton @hannes-steffenhagen-diffblue added a unit test and adressed comments

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a76b03d49ab45cc300a881bb1b1d6de4ae5d49bb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85444852
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: d13a73f215e762c3f2d58f03670d2728a601ca44).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85447907

smowton reviewed (approved):
Excellent, thankyou very much for doing the extra work!

tautschnig commented:
I really like how this one evolved from a commit message that I wasn't able to decipher to a PR that even I understand.

mgudemann commented:
@tautschnig if you didn't understand, then we'd very likely have a problem!

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 5a0feaa1a7278d25d3730b5994eaa94a170668a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85471472

peterschrammel reviewed (approved):
LGTM
